### PR TITLE
Fix data access races when modifying finished spans as they're being flushed

### PR DESCRIPTION
**Pull request by elijahandrews**
We don't lock spans before flushing them. If a user finishes a span, then updates the same span, we have a data race. I've added a test which demonstrates this; it should be run with `go test -race`.

Though it doesn't make a lot of sense to change a span after it's been added to the buffer, there's nothing preventing the user from doing this. We should be defensive about this, as newer versions of go crash on concurrent map accesses, and we store metadata and metrics in maps.

### CODE REVIEW

**Comment by ethanwp** [comment]
Locking would cause strange contention as I assume the network write would presumably block and hold up the lock for a long time? I would suggest after calling .Finish(), additional write methods via the API should simply do nothing.

**Comment by elijahandrews** [comment]
Good point, we will be holding these locks over `SendTraces`, which may be a network call depending on the transport. As you suggested, another way I considered fixing this was to update the `Set*` methods to ignore updating the spans if a span is finished. However, I think my solution is better for two reasons:

1. I don't expect much contention on spans after they've been finished, as during correct client usage I don't see why this would happen; this race only happened when there was a bug in my tracing code. If we add checks in the setters, we're adding an extra check for typical client usage, while here we're only adding contention during incorrect usage. I'd rather add overhead the uncommon, incorrect case than the typical case.
2. As we add more functionality to `dd-trace-go`, we may forget to add this check to the new code. Instead of implicitly locking in our setters, I think it makes more sense to use a real lock.

I suppose we could also avoid holding the lock over the network call by copying the spans and then unlocking them before the network call. However, this solution also adds overhead to the typical usage case, which I'd like to avoid.

On the other hand, blocking a thread calling a span setter on a network call could be disastrous, depending on how the library is used. I'll defer this decision to the repo maintainers.

**Comment by elijahandrews** [comment]
@palazzem I was writing this PR under the assumption that Spans are thread-safe, but now that I look closer, maybe that isn't the case. Are they supposed to be? If so, shouldn't we create thread-safe getters and setters for all of its attributes, then make its attributes private?

**Comment by elijahandrews** [comment]
I should note that I'm particularly worried about data races in `span.Meta` and `span.Metrics`, because these races cause more than just correctness issues. In Go 1.6 and later, concurrent map reads and writes will panic, so we may crash programs that expose the race that I've fixed in this PR.

**Comment by palazzem** [comment]
@elijahandrews that's exactly my concern. Your PR raises an issue that must be addressed as soon as possible. I will think a little about it and will be back in this thread 👍

**Comment by elijahandrews** [comment]
Any update on this?

**Comment by ufoot** [comment]
Taking a look at this, I think @elijahandrews your initial idea of checking wether a Span is finished within setters, and discard them it it's indeed finished, is good after all. We could even at some point record (log message, metric report, hint to agent, whatever is relevant) the fact some Span changing function has been called *AFTER* it's finished. This is a programming error and probably a good way to avoid it is to say it aloud when we notice rather than relying on people reading the docs...  We really do not want users to do this, we can technically get rid of the race condition, but I'd like as well to get rid of misunderstandings of the way the lib works -> if the setters are disabled, it should at some point be quite obvious for users that something is going wrong (non-working code + log message = increased probability it's going to be updated at some point). I note you would prefer not to add a check ("is finished ?") to the main code path and indeed it's wise to keep that one as Slack as possible, but yet I think we can afford an extra boolean check here.

This won't  solve the `span.Service = "a wrong change?"` issue @palazzem described but since all data fields are public in the `Span` struct this is really hard to avoid, and you can reasonably expect accessing direct fields *not* to be thread-safe. One could even write `span.Meta["key"]="an even more wrong change?"` for that matter (and that is dangerous for sure).

As a side note, we might want to optimise locking here (an RWMutex could help in GetMeta)

File: tracer/span.go [hunk 1]
```diff
@@ -136,7 +152,20 @@ func (s *Span) SetError(err error) {
 		return
 	}
 
+	// FIXME: We acquire and release the lock several times over the course of
```
Comment (elijahandrews) [inline tracer/span.go]: At some point you should do a pass where you replace the span's lock with an `RWMutex`, and redo the way locking works so you only need to acquire the lock once in this function. I think it's out of scope to fix that in this PR.

**Comment by elijahandrews** [comment]
@ufoot thanks for the comment. I've modified the PR with the changes you've requested. I'll also open an issue about the public data fields on the `Span` that cannot be safely accessed; I really think we should make them private and write accessors for them.

**Comment by elijahandrews** [comment]
CI is failing, but it appears to be unrelated to my change:

```
go get -t ./tracer/...
package gopkg.in/go-playground/validator.v8: unrecognized import path "gopkg.in/go-playground/validator.v8" (https fetch: Get https://gopkg.in/go-playground/validator.v8?go-get=1: remote error: tls: internal error)
package gopkg.in/yaml.v2: unrecognized import path "gopkg.in/yaml.v2" (https fetch: Get https://gopkg.in/yaml.v2?go-get=1: remote error: tls: internal error)
rake aborted!
Command failed with status (1): [go get -t ./tracer/......]
/home/ubuntu/.go_workspace/src/github.com/DataDog/dd-trace-go/Rakefile:46:in `block in <top (required)>'
Tasks: TOP => ci => get
(See full trace by running task with --trace)

cd "$IMPORT_PATH" && rake ci returned exit code 1
```

Will try running it again tomorrow morning and seeing if the problem is fixed.

**Comment by elijahandrews** [comment]
Tests pass now! Ready for another review pass.

File: tracer/span.go [hunk 2]
```diff
@@ -121,12 +141,20 @@ func (s *Span) SetMetric(key string, val float64) {
 		return
 	}
 
-	s.mu.Lock()
+	s.Lock()
+	defer s.Unlock()
+
+	// We don't lock spans when flushing, so we could have a data race when
+	// modifying a span as it's being flushed. This protects us against that
+	// race, since spans are marked `finished` before we flush them.
```
Comment (ufoot) [inline tracer/span.go]: I think that comment, here, is not really relevant as, indeed, it *IS* locked in that place.

File: tracer/span.go [hunk 3]
```diff
@@ -58,9 +58,9 @@ type Span struct {
 	Error    int32              `json:"error"`             // error status of the span; 0 means no errors
 	Sampled  bool               `json:"-"`                 // if this span is sampled (and should be kept/recorded) or not
 
-	tracer   *Tracer    // the tracer that generated this span
-	mu       sync.Mutex // lock the Span to make it thread-safe
-	finished bool       // true if the span has been submitted to a tracer.
+	sync.RWMutex
```
Comment (ufoot) [inline tracer/span.go]: Here we're slighly modifying our API and exposing the RWMutex methods AFAIK. Not a big big deal but will probably require another pair of eyes as this does change what we expose to users. I, for one, am fine with it.

Comment (elijahandrews) [inline tracer/span.go]: Yup, indeed this is the case. I think it makes sense to make this lock public, as there was previously no thread-safe way to access many of the public fields :)

File: tracer/span.go (hunk 2 above)
Comment (elijahandrews) [inline tracer/span.go]: What I'm trying to say with this comment is that even though we've locked the span here, it still isn't safe to modify `s.Metrics` since the flushing code doesn't lock the spans. I can update these comments to be a bit clearer.

Comment (elijahandrews) [inline tracer/span.go]: on second thought, I think I'll leave the comment as-is.

**Comment by elijahandrews** [comment]
Bump to get this merged, it bit us again today.

**Comment by LotharSee** [comment]
Thanks for the catch and the fix!

While I'm good with the code, my main concern here is that it's a breaking change (even if legit) for anyone modifying a span after a finish.

While I agree it is globally a good practice to have `Finish` not allowing future modification, we should:

- Update the doc to explicitly state that one can't modify a span after finishing it
- Add the ability to set the end/duration without really closing the span (like `span.SetEnd()`). Legit use-case:

```go
func Whatever() {
  span := tracer.NewRootSpan("worker.process", "bg-worker", "legit-business")
  // Run some quite complicated equations
  // Done with the computing!
  
  // Now I do a non-trivial operation required to set one of my tags
  notification := Notifier.send_my_computing_result()
  span.set_tag("notification.id", notification.ID)
  
  // If I were to do span.Finish() now, that would contain the time of whatever I needed to run to set my tags
  // I should be able to set my duration after "Done with the computing!"
}
```

The second point can wait another PR, until we internally decide what we want in terms of API (and we should maybe port a similar change to our other clients?).



TL;DR: let's update the doc to say `Finish` doesn't allow future change.

**Comment by elijahandrews** [comment]
Docs updated as requested.

**Comment by LotharSee** [comment]
GTM!
